File: src/lib/manifest/types/stores.ts
/**
 * Shapes of the derived entity stores persisted as JSON under
 * `.cache/entities/<manifestVersion>/`. One file per store. Each store is the
 * output of exactly one extractor and is rebuilt only when the manifest
 * version changes.
 */

import type {
  AbilityRecord,
  ArtifactRecord,
  AspectRecord,
  ExoticArmorRecord,
  ExoticWeaponRecord,
  FragmentRecord,
  Hash,
  ModRecord,
  OriginTraitRecord,
  PerkRecord,
  SetBonusRecord,
  WeaponRecord,
} from "./records";

export interface EntityStores {
  "exotic-armor": ExoticArmorRecord[];
  "exotic-weapons": ExoticWeaponRecord[];
  weapons: WeaponRecord[];
  /** All perks referenced by weapon perk columns, keyed for joins. */
  "weapon-perks": PerkRecord[];
  "origin-traits": OriginTraitRecord[];
  artifacts: ArtifactRecord[];
  aspects: AspectRecord[];
  fragments: FragmentRecord[];
  abilities: AbilityRecord[];
  mods: ModRecord[];
  "set-bonuses": SetBonusRecord[];
  /** Stat display names from DestinyStatDefinition, keyed by hash. */
  stats: { hash: Hash; name: string; description: string }[];
}

export type StoreName = keyof EntityStores;

export const STORE_NAMES: readonly StoreName[] = [
  "exotic-armor",
  "exotic-weapons",
  "weapons",
  "weapon-perks",
  "origin-traits",
  "artifacts",
  "aspects",
  "fragments",
  "abilities",
  "mods",
  "set-bonuses",
  "stats",
] as const;

/** Metadata sidecar written alongside the stores. */
export interface EntityCacheMeta {
  manifestVersion: string;
  builtAt: string;
  counts: Record<StoreName, number>;
}

File: packages/manifest/lib/src/types/stores.dart
import 'records.dart';

/// MVP entity store names (DART-017). Product has a larger set; later slices expand.
enum MvpStoreName {
  weapons('weapons'),
  exoticArmor('exotic-armor'),
  aspects('aspects'),
  fragments('fragments'),
  abilities('abilities'),
  mods('mods');

  const MvpStoreName(this.fileStem);
  final String fileStem;

  static MvpStoreName? tryParse(String value) {
    for (final v in values) {
      if (v.fileStem == value || v.name == value) return v;
    }
    return null;
  }

  static const List<MvpStoreName> all = values;
}

/// Metadata sidecar written alongside the stores.
class EntityCacheMeta {
  const EntityCacheMeta({
    required this.manifestVersion,
    required this.builtAt,
    required this.counts,
  });

  final String manifestVersion;
  final String builtAt;
  final Map<String, int> counts;

  Map<String, dynamic> toJson() => {
        'manifestVersion': manifestVersion,
        'builtAt': builtAt,
        'counts': counts,
      };

  factory EntityCacheMeta.fromJson(Map<String, dynamic> json) {
    final rawCounts = json['counts'];
    final counts = <String, int>{};
    if (rawCounts is Map) {
      for (final e in rawCounts.entries) {
        counts[e.key.toString()] = (e.value as num).toInt();
      }
    }
    return EntityCacheMeta(
      manifestVersion: json['manifestVersion'] as String? ?? '',
      builtAt: json['builtAt'] as String? ?? '',
      counts: counts,
    );
  }
}

/// Typed decode for MVP stores from JSON list.
List<T> decodeStoreList<T>(
  MvpStoreName store,
  List<dynamic> json,
  T Function(Map<String, dynamic>) fromJson,
) {
  return json
      .map((e) => fromJson(Map<String, dynamic>.from(e as Map)))
      .toList();
}

List<dynamic> encodeStoreRecords(MvpStoreName store, List<Object> records) {
  return records.map((r) {
    if (r is ExoticArmorRecord) return r.toJson();
    if (r is WeaponRecord) return r.toJson();
    if (r is AspectRecord) return r.toJson();
    if (r is FragmentRecord) return r.toJson();
    if (r is AbilityRecord) return r.toJson();
    if (r is ModRecord) return r.toJson();
    throw ArgumentError('Unknown record type for $store: ${r.runtimeType}');
  }).toList();
}
